opm/autodiff/FullyImplicitBlackoilSolver_impl.hpp

```diff
@@ -1896,83 +1923,55 @@ namespace {
+        std::array<double,MaxNumPhases> CNV                   = {{0., 0., 0.}};
+        std::array<double,MaxNumPhases> R_sum                 = {{0., 0., 0.}};
+        std::array<double,MaxNumPhases> B_avg                 = {{0., 0., 0.}};
+        std::array<double,MaxNumPhases> maxCoeff              = {{0., 0., 0.}};
+        std::array<double,MaxNumPhases> mass_balance_residual = {{0., 0., 0.}};
+        std::size_t cols = MaxNumPhases; // needed to pass the correct type to Eigen
+        Eigen::Array<V::Scalar, Eigen::Dynamic, MaxNumPhases> B(nc, cols);
+        Eigen::Array<V::Scalar, Eigen::Dynamic, MaxNumPhases> R(nc, cols);
+        Eigen::Array<V::Scalar, Eigen::Dynamic, MaxNumPhases> tempV(nc, cols);
+
+        for(int idx=0; idx<MaxNumPhases; ++idx)
+        {
+            if (active_[idx]) {
+                const int pos = pu.phase_pos[idx];
+                const ADB& tempB = rq_[pos].b;
+                B.col(pos)  = 1./tempB.value();
+                R.col(pos) = residual_.material_balance_eq[idx].value();
+                tempV.col(pos) = R.col(pos).abs()/pv;
+            }
         }
 
-        if (active_[Gas]) {
-            // Omit the vapoil here. We should add it.
-            const int pos = pu.phase_pos[Gas];
-            const ADB& tempBG = rq_[pos].b;
-            V BG = 1./tempBG.value();
-            V RG = residual_.material_balance_eq[Gas].value();
-            BG_avg = BG.sum()/nc;
-            const V tempV = RG.abs()/pv;
+        const double pvSum = convergenceReduction(B, tempV, R, R_sum, maxCoeff, B_avg, nc);
 
-            CNVG = BG_avg * dt * tempV.maxCoeff();
-            RG_sum = RG.sum();
+        bool converged_MB = true;
+        bool converged_CNV = true;
+        // Finish computation
+        for(int idx=0; idx<MaxNumPhases; ++idx)
+        {
+            CNV[idx] = B_avg[pu.phase_pos[idx]] * dt * maxCoeff[pu.phase_pos[idx]];
+            mass_balance_residual[idx] = std::abs(B_avg[pu.phase_pos[idx]]*R_sum[pu.phase_pos[idx]]) * dt / pvSum;
+            converged_MB = converged_MB && (mass_balance_residual[idx] < tol_mb);
+            converged_CNV = converged_CNV && (CNV[idx] < tol_cnv);
         }
 
-        const double mass_balance_residual_water = std::abs(BW_avg*RW_sum) * dt / pvSum;
-        const double mass_balance_residual_oil   = std::abs(BO_avg*RO_sum) * dt / pvSum;
-        const double mass_balance_residual_gas   = std::abs(BG_avg*RG_sum) * dt / pvSum;
-
-        bool converged_MB = (mass_balance_residual_water < tol_mb)
-                         && (mass_balance_residual_oil   < tol_mb)
-                         && (mass_balance_residual_gas   < tol_mb);
-
-        bool converged_CNV = (CNVW < tol_cnv) && (CNVO < tol_cnv) && (CNVG < tol_cnv);
-
-        const double residualWellFlux = infinityNorm( residual_.well_flux_eq );
-        const double residualWell     = infinityNorm( residual_.well_eq );
-
+        double residualWellFlux = infinityNorm(residual_.well_flux_eq);
```

Comment:
This line and the next seems to be a small step backwards: the const is removed and it is no longer formatted nicely aligned, is this just an accidental edit?